modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_1_0/UpgradeDDMStructure.java

```diff
@@ -230,6 +231,45 @@ private String _upgradeDDMFormLayoutDefinition(
 		return ddmFormLayoutSerializerSerializeResponse.getContent();
 	}
 
+	private String _upgradeDDMFormLayoutDefinition(
+			String structureLayoutDefinition, String structureVersionDefinition)
+		throws Exception {
+
+		DDMFormLayout ddmFormLayout = DDMFormLayoutDeserializeUtil.deserialize(
+			_ddmFormLayoutDeserializer, structureLayoutDefinition);
+
+		DDMForm ddmForm = DDMFormDeserializeUtil.deserialize(
+			_ddmFormDeserializer, structureVersionDefinition);
+
+		DDMFormLayoutPage ddmFormLayoutPage =
+			ddmFormLayout.getDDMFormLayoutPage(0);
+
+		List<DDMFormLayoutRow> ddmFormLayoutRowList =
+			ddmFormLayoutPage.getDDMFormLayoutRows();
+
+		DDMFormLayoutRow ddmFormLayoutRow = new DDMFormLayoutRow();
+
+		for (DDMFormField ddmFormField : ddmForm.getDDMFormFields()) {
+			ddmFormLayoutRow.addDDMFormLayoutColumn(
```

Comment:
Do you think that it would be better to create a row for each field? I know that OLD DDM does not allow more than one field per line, but the upgraded structure can be edited in the DE Taglib, which allows more than one field per line.
Is ddmForm.getDDMFormFields()  returning the ordered list?

Reply:
Thanks you very much for your comment, I think it fixed the upgrade issues, :). 

Regarding the ordered list, I don't know, I did some tests and it was working, I can add some test to check it if you want.